model/fn-execution/src/main/proto/org/apache/beam/model/fn_execution/v1/beam_fn_api.proto

```diff
@@ -1075,6 +1111,42 @@ message StateClearRequest {}
 // A response to clear state.
 message StateClearResponse {}
```

Comment:
For symmetry, its seems we should be re-using the Get/Append/Clear messages just like with all the other states, and stick a range in the state key (for get and clear at least; for append it'd be the sort keys provided in the data). Otherwise ordered list state is structured differently from all the rest.

Reply:
There is a discussion for this in the original design doc. https://docs.google.com/document/d/1U77sAvE6Iy9XsVruRYHxPdFji7nqS6HPi1XU8fhyrxs/edit?disco=AAAAH0MsNik

I think @kennknowles mentioned he preferred separate messages.

Personally, I prefer the way you propose to re-use the Get/Append/Clear messages by putting optional fields there for ordered list state. 

However, I hesitate to do that because (1) it will introduce new optional fields to an existing proto message and (2) regarding extensibility and flexibility, it may be easier to have separate messages for ordered list and any future states that doesn't fit in current Get/Append/Clear paradigm?

WDYT?

Reply:
I prefer being consistent. If we want to restructure things to break everything out, we can have that discussion, but it doesn't seem a clear enough win to break consistency. (If the range is in the ordered list state state key, the existing Get/Append/Clear will work just fine, no need for optional fields.)

@kennknowles for your thoughts too.

Reply:
I still agree with my comment there: "it looks like a 'do everything' request/response protocol and the key holds all the information." That comment may seem to be about style, but it is really about transparency, debuggability, readability.

I view whatever "consistency" we have amongst the state requests as coincidence at best and a mistake at worst. _The_ reason that different kinds of state exist is because they support different methods. There is a reason that the only method they have in common in Java (where it originated) is `clear()`. Apparently the design decision was to model `methodA, methodB, methodC` as `oneBigMethod("actually do A"), oneBigMethod("actually do B"), oneBigMethod("actually do C")`. I just think that is not as good as being straightforward.

Reply:
The one counterpoint is that a state key is used for caching, as long as the spec is that a state key is deterministic in what it returns. So it has value for `get`.

Reply:
OK my responses are evolving now that I've read the whole code change and re-read the doc.

 - I see specialized requests are broken out anyhow, so that's fine.
 - Including everything needed for caching in the state key is good for raw request caching, so re-using Get is good. Though there are perhaps smarter ways to cache that won't benefit from this
 - "append" is still a fine method for adding things to ordered list state, but it isn't important and the name is misleading (as it is for bags and multimaps, since they are not ordered, so anyhow it is the same here and might as well keep the incorrect naming)
 - Obviously `clear` is fine

Reply:
Sounds good! Thanks a lot for the input, Kenn.

I will go ahead with the changes to re-use Get/Append/Clear for ordered list then.

Reply:
Thanks. 

IIUC, append is still correct, as there's a bag (with possibly multiple items) assigned to every point in the ordered space. (It's a lot like MultiMap with the ability to read ranges in order rather than just do point lookups, though +1 to not mixing the two.) 

Agree on caching--there are more clever things we can do here in the future, but we can punt that to future work.

Reply:
I pushed the new code to reuse Get/Append/Clear.